lib/common/rules.c

```diff
@@ -231,3 +231,112 @@ pcmk__evaluate_date_spec(const xmlNode *date_spec, const crm_time_t *now)
     // All specified ranges passed, or none were given (also considered a pass)
     return pcmk_rc_ok;
 }
+
+/*!
+ * \internal
+ * \brief Given a duration and a start time, calculate the end time
+ *
+ * \param[in]  duration  XML of PCMK_XE_DURATION element
+ * \param[in]  start     Start time
+ * \param[out] end       Where to store end time (must contain NULL initially)
+ *
+ * \return Standard Pacemaker return code
+ * \note The caller is responsible for freeing \p *end using crm_time_free().
+ */
+int
+pcmk__unpack_duration(const xmlNode *duration, const crm_time_t *start,
+                      crm_time_t **end)
+{
+    int rc = pcmk_rc_ok;
+    int sub_rc = pcmk_rc_ok;
+    const char *id = NULL;
+
+    if ((start == NULL) || (duration == NULL)
+        || (end == NULL) || (*end != NULL)) {
+        return EINVAL;
+    }
+
+    // Get duration ID (for logging)
+    id = pcmk__xe_id(duration);
+    if (pcmk__str_empty(id)) { // Not possible with schema validation enabled
+        /* @COMPAT When we can break behavioral backward compatibility,
+         * return pcmk_rc_unpack_error instead
+         */
+        pcmk__config_warn(PCMK_XE_DURATION " element has no " PCMK_XA_ID);
+        id = "without ID";
+    }
+
+    *end = pcmk_copy_time(start);
+
+    sub_rc = pcmk__add_time_from_xml(*end, pcmk__time_years, duration);
```

Comment:
I'd prefer an inline or macro to check the RC and warn. Probably should pair that with an array or switch mapping the enum values to their string equivalents. Maybe in a `pcmk__time_text()` helper. This is a lot of duplication.

Reply:
Yeah, this eventually simplifies but probably better to come up with something now